prysmaticlabs/prysm
# Spec checker tool

**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**
- This tool allows to download specs documentation, parse Python code blocks from it, and then use parsed blocks to compare with documentation pseudo code we are using when implementing specs functions. Otherwise, it is a simple CLI utility relying on AST parsing.
- The delta of this PR is big, due to inclusion of parsed reference specs (see the rationale for this below).
- This PR implements the separate CLI tool (and not the static analyzer), because `go_tool_library` doesn't support `embededsrcs` attribute (which `go_library` does support), and our tool relies on `go:embed`. Once `rules_go` are updated, we can turn this tool into static analyzer, easily.
- Multiple definitions with the same function names are possible (for example you can have `def func_name():` defined for both phase1 and phase0). As long as one of the definitions matches our comments, we consider comments to be up to date.

**Which issues(s) does this PR fix?**

Fixes #5717

**How to update referenced specs?**
```bash
bazel run //tools/specs-checker download -- --dir=$PWD/tools/specs-checker/data
```
This will pull the files defined in `specDirs` (see `main.go`), parse them (extract Python code snippets, discarding any other text), and save them to the folder from which `bazel run //tools/specs-checker check` will be able to embed (using `go:embed`):

![image](https://user-images.githubusercontent.com/188194/114362104-a36e5180-9b7f-11eb-939c-790a17ccc3bc.png)

Note: see included README for more details.

**How to run tests?**
```bash
bazel run //tools/specs-checker check -- --dir $PWD/beacon-chain
bazel run //tools/specs-checker check -- --dir $PWD/validator
bazel run //tools/specs-checker check -- --dir $PWD/shared
```
Or, to check the whole project:
```bash
bazel run //tools/specs-checker check -- --dir $PWD
```

**How do results of the check look like?**
![image](https://user-images.githubusercontent.com/188194/114014204-05be0e00-9871-11eb-87a2-77a13d4e37d2.png)

**Why save parsed reference definitions, why not discard them after the check?**
There are several reasons for doing that:
- I found it particularly useful when downloading *updates*, since reference definitions are versioned, on `git diff` I immediately see what files have changed, to better understand the scope of updates I need to do on our comments.
- When used as static analyzer, re-downloading from GitHub on every build attempt is not the best idea, so having those results cached comes handy. I could have implemented more sophisticated caching scheme, where tool will check the update time of the spec files and re-download only if older than a given period, but I consider it as a bit of an overkill for the tool that simple. So, it is up to us to decide how often do we want to pull updated specs from the upstream.
- Finally, it is way easier to debug, as you have all the reference definitions saved within our project. As a side-kick: when updating the comments (if it is out of date), you can just grep the saved reference docs, and do not have to fish through specs repo site in order to find the up to date version of a snippet.
